module/component/component.go

```diff
@@ -174,34 +201,112 @@ func (c *componentManagerBuilderImpl) AddWorker(worker ComponentWorker) Componen
 	return c
 }
 
-func (c *componentManagerBuilderImpl) Build() *ComponentManager {
-	return &ComponentManager{
-		started: atomic.NewBool(false),
-		ready:   make(chan struct{}),
-		done:    make(chan struct{}),
-		workers: c.workers,
+func (c *componentManagerBuilderImpl) Build() ComponentManager {
+	shutdownSignal := make(chan struct{})
+
+	taskRunner := &taskRunner{
+		tasks:          make(chan *runTaskRequest),
+		asyncTasks:     make(chan ComponentTask),
+		shutdownSignal: shutdownSignal,
+	}
+
+	// copy workers to a new slice to prevent further modifications
+	workers := make([]ComponentWorker, len(c.workers)+1)
+	workers[0] = taskRunner.handleTasks
+	copy(workers[1:], c.workers)
+
+	return &componentManagerImpl{
+		started:        atomic.NewBool(false),
+		ready:          make(chan struct{}),
+		done:           make(chan struct{}),
+		shutdownSignal: shutdownSignal,
+		workers:        workers,
+		taskRunner:     taskRunner,
+	}
+}
+
+var ErrComponentManagerClosed = errors.New("component manager is closed")
+
+type taskRunner struct {
+	tasks          chan *runTaskRequest
+	asyncTasks     chan ComponentTask
+	shutdownSignal chan struct{}
+}
+
+type runTaskRequest struct {
+	task ComponentTask
+	done chan struct{}
+}
+
+func (t *taskRunner) Run(task ComponentTask) error {
+	done := make(chan struct{})
+
+	select {
+	case <-t.shutdownSignal:
+		return ErrComponentManagerClosed
+	case t.tasks <- &runTaskRequest{task, done}:
+		select {
+		case <-t.shutdownSignal:
+			return ErrComponentManagerClosed
+		case <-done:
+			return nil
+		}
+	}
+}
+
+func (t *taskRunner) RunAsync(task ComponentTask) {
```

Comment:
I would point out that according to https://github.com/onflow/flow/blob/master/flips/20210201-consuming-messages-from-network-layer.md, we eventually want to move towards all Engines having a queue for messages which get consumed by a separate routine.

The approach implemented here is consistent with this, and it would be easy to modify things to add this functionality.

Reply:
At this moment almost all of our engines(at least consensus) are using queening mechanism and they are implemented in a way that `Do` is more practical for this operation. We specifically designed queues in a way that insertion is not blocking and can be executed on network layer stack.